single_kernel_mongo/managers/mongo.py

```diff
@@ -0,0 +1,415 @@
+
+    def initialise_monitor_user(self):
+        """Creates the monitor user on the MongoDB database."""
+        if self.state.app_peer_data.is_user_created(MonitorUser.username):
+            return
+        with MongoConnection(self.state.mongo_config) as mongo:
+            logger.debug("Creating the monitor user roles…")
+            mongo.create_role(
+                role_name=MonitorUser.mongodb_role,
+                privileges=MonitorUser.privileges,
+            )
+            logger.debug("creating the monitor user...")
+            mongo.create_user(self.state.monitor_config)
+        self.state.app_peer_data.set_user_created(MonitorUser.username)
+
+    def initialise_backup_user(self):
+        """Creates the monitor user on the MongoDB database."""
+        if self.state.app_peer_data.is_user_created(BackupUser.username):
+            return
+        with MongoConnection(self.state.mongo_config) as mongo:
+            logger.debug("Creating the backup user roles…")
+            mongo.create_role(
+                role_name=BackupUser.mongodb_role,
+                privileges=BackupUser.privileges,
+            )
+            logger.debug("creating the backup user...")
+            mongo.create_user(self.state.backup_config)
+        self.state.app_peer_data.set_user_created(BackupUser.username)
+
+    def oversee_relation(
+        self, relation: Relation, relation_departing: bool = False, relation_changed: bool = False
+    ):
+        """Oversees the users of the relation.
+
+        Function manages user by removing, updated, and creating
+        users; and dropping databases when necessary.
+
+        Args:
+            relation: The relation are working with.
+            departing: If this is a relation broken event.
+            event: relation event or None.
+
+        Raises:
+            PyMongoError
+        """
+        self.add_user(relation)
+        self.update_user(relation)
+        if relation_departing:
+            self.remove_user(relation)
+            self.auto_delete_db(relation)
+        if relation_changed:
+            self.update_diff(relation)
+
+    def update_diff(self, relation: Relation):
+        """Update the relation databag with the diff of data.
+
+        Args:
+            relation: The relation to update the databag from.
+        """
+        if not self.charm.unit.is_leader():
+            return
+        data_interface = DatabaseProviderData(
+            self.model,
+            relation.name,
+        )
+        actual_data = data_interface.fetch_relation_data([relation.id]).get(relation.id, {})
+        new_data = {key: value for key, value in actual_data.items() if key != "data"}
+        data_interface.update_relation_data(relation.id, {"data": json.dumps(new_data)})
+
+    def add_user(self, relation: Relation):
+        """Add a user for this relation."""
+        managed_users = self.state.app_peer_data.managed_users
```

Comment:
Yes but : as much as possible I try to get the initialisation of the variables on top of the function body: the idea behind that is that we don't end up trying to use variables before they are initialised, and since getting this data is not a heavy and expensive process, better add it early.
However, while reading this code again, I noticed that there is another variable that should be moved on top so I'll move it :)